tools/src/main/scala/scala/scalanative/build/Discover.scala

```diff
@@ -12,14 +12,94 @@ import scalanative.build.IO.RichPath
+  /** List of source patterns used */
+  val jarExtension  = ".jar"
+  val srcExtensions = Seq(".c", ".cpp", ".S")
+  val srcPatterns   = srcExtensions.mkString("glob:**{", ",", "}")
+
+  /** Original jar or dir path and generated dir path for native code */
+  case class NativeLib(src: Path, dest: Path)
+
+  /** To positively identify nativelib */
+  val nativeLibMarkerFile = "org_scala-native_nativelib.txt"
+
+  private lazy val dirMarkerFilePattern = "glob:**" + nativeLibMarkerFile
+
+  def isJar(path: Path): Boolean = path.toString().endsWith(jarExtension)
+
+  object NativeLib {
+    def isJar(nativelib: NativeLib): Boolean = Discover.isJar(nativelib.src)
+  }
+
+  /** Compilation mode name from SCALANATIVE_MODE env var or default. */
   def mode(): String =
     getenv("SCALANATIVE_MODE").getOrElse(build.Mode.default.name)
 
-  /** LTO variant used for release mode. */
+  /** LTO variant used for release mode from SCALANATIVE_LTO env var or default. */
   def LTO(): String =
     getenv("SCALANATIVE_LTO").getOrElse("none")
 
+  /** GC variant used from SCALANATIVE_GC env var or default. */
+  def GC(): String =
+    getenv("SCALANATIVE_GC").getOrElse(build.GC.default.name)
+
+  private[build] def findNativeLibs(classpath: Seq[Path],
+                                    workdir: Path): Seq[NativeLib] = {
+    val nativeLibPaths = classpath.flatMap { path =>
+      if (isJar(path)) readJar(path)
+      else readDir(path)
+    }
+
+    val extractPaths =
+      for ((path, index) <- nativeLibPaths.zipWithIndex) yield {
+        val name =
+          path
+            .getName(path.getNameCount() - 1)
+            .toString()
+            .stripSuffix(jarExtension)
+        NativeLib(src = path,
+                  dest = workdir.resolve(s"native-code-$name-$index"))
+      }
+
+    if (extractPaths.isEmpty)
+      throw new BuildException(s"No Native Libraries found: $classpath")
+    else
+      extractPaths
+  }
+
+  private[build] def findNativeLib(nativeLibs: Seq[NativeLib]): Path = {
+    val nativeLib = nativeLibs.find { nl =>
+      val srcPath = nl.src
+      if (isJar(srcPath))
+        IO.existsInJar(srcPath, hasMarkerFileInJar)
+      else
+        IO.existsInDir(srcPath, dirMarkerFilePattern)
+    }
+    nativeLib match {
+      case Some(nl) => nl.dest
+      case None =>
+        throw new BuildException(s"Native Library not found: $nativeLibs")
+    }
+  }
+
+  private def isNativeFile(name: String): Boolean =
+    srcExtensions.map(name.endsWith(_)).exists(identity)
+
+  private def hasMarkerFileInJar(name: String): Boolean =
+    name.endsWith(nativeLibMarkerFile)
+
+  private def readDir(path: Path): Option[Path] =
+    IO.existsInDir(path, srcPatterns) match {
+      case true  => Some(path)
+      case false => None
+    }
+
+  private def readJar(path: Path): Option[Path] =
+    IO.existsInJar(path, isNativeFile) match {
+      case true  => Some(path)
+      case false => None
+    }
```

Comment:
Thinking more about this, I think we should restrict the search for native code in a specific directory within the Dir/Jar. For example inside `META-INF/scala-native-native-libs/`.

Otherwise, codebases will never be able to put `.c`, `.cpp` or `.S` files in their `resources` without them being considered by the Scala Native linker as native libs that it should necessarily try to compile and link. That's not a great user experience.

Reply:
I am not super excited about that idea after reading online about how `META-INF` is used.

A couple of thoughts.

1. Our only customers are Scala library authors wishing to make their library work with Scala Native.
2. What are the chances that they are putting these files in the resource directory and for what purpose?

We could decide on a special directory inside resources but then we would have change the current code. Nothing we could come up with is 100% foolproof and would restrict library authors so I am not sure it is worth the effort for a hypothetical. 

That said, we could internally migrate code to `resources/native` and recommend that to library authors so that if we ever have a conflict we could change the code to use that directory only. It looks to me like it would take another PR to make the Scala Native `nativelib` compile if the code was moved into `resources/native`. Any other library with native code shouldn't matter. I have tested this with the scripted test and have documentation prepared if you agree.

Reply:
Just as a data point searching all of Github. `"src/resources/native" in:path`
You can also add the that `langauge:java` or Scala and see even less.

It is used but very little and no source code matches I could find. If we wanted to be super safe we could go with `scala-native` which is more specific but not too ugly.

Reply:
If we went with `resources/scala-native` we could do a followup PR to enforce it.

Reply:
> What are the chances that they are putting these files in the resource directory and for what purpose?

It's not a question of "chances". If there is a chance, we need to take it into account. In language design, you have to account for *all* possible cases, otherwise you will block *some*one, somewhere.

> If we wanted to be super safe we could go with scala-native which is more specific but not too ugly.

`scala-native` sounds good.